subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/adapter/ProtocolToModelAdapter.java

```diff
@@ -631,16 +632,30 @@ public void invoke(MethodInvocation method) throws Throwable {
                 Object value;
                 next.invoke(method);
                 if (!method.found()) {
-                    unknown.add(method.getName());
+                    markUnknown(method.getName());
                     return;
                 }
                 value = method.getResult();
-                properties.put(method.getName(), value);
+                cachePropertyValue(method.getName(), value);
                 return;
             }
 
             next.invoke(method);
         }
+
+        private void markUnknown(String methodName) {
+            if (unknown.isEmpty()) {
+                unknown = new HashSet<String>();
+            }
+            unknown.add(methodName);
```

Comment:
aren't these initialized fully on first use anyway?

Reply:
If no unknown methods are called, this collection is never initialized